ironfish/src/network/peerNetwork.ts

```diff
@@ -657,24 +719,278 @@ export class PeerNetwork {
+    let hashesConsumed = 0
+    let fullTransactionsConsumed = 0
+    let nextFullTransaction = absoluteIndexTransactions.next()
+
+    const partialTransactions: TransactionOrHash[] = []
+    const missingTransactions: number[] = []
+
+    while (hashesConsumed < numHashes || !nextFullTransaction.done) {
+      const currPosition = hashesConsumed + fullTransactionsConsumed
+
+      // If we have no more full transactions or a transaction doesn't belong in this position
+      if (nextFullTransaction.done || currPosition !== nextFullTransaction.value.index) {
+        if (hashesConsumed === numHashes) {
+          // We ran out of hashes to populate
+          return { ok: false }
+        }
+
+        const hash = block.transactionHashes[hashesConsumed]
+        const transaction = this.node.memPool.get(hash)
+        const resolved: TransactionOrHash = transaction
+          ? {
+              type: 'FULL',
+              value: transaction,
+            }
+          : {
+              type: 'HASH',
+              value: hash,
+            }
+        if (resolved.type === 'HASH') {
+          missingTransactions.push(currPosition)
+        }
+
+        partialTransactions.push(resolved)
+        hashesConsumed++
+        continue
       }
+
+      partialTransactions.push({
+        type: 'FULL',
+        value: new Transaction(nextFullTransaction.value.transaction),
+      })
+      nextFullTransaction = absoluteIndexTransactions.next()
+      fullTransactionsConsumed++
     }
+
+    return { ok: true, partialTransactions, missingTransactions }
   }
 
-  private handleNewBlockHashesMessage(peer: Peer, message: NewBlockHashesMessage) {
-    this.logger.debug(`Received unimplemented message ${message.type}`)
+  assembleBlockFromResponse(
+    partialTransactions: TransactionOrHash[],
+    responseTransactions: readonly SerializedTransaction[],
+  ): { ok: false } | { ok: true; transactions: Transaction[] } {
+    const transactions: Transaction[] = []
+    let currResponseIndex = 0
+
+    for (const partial of partialTransactions) {
+      if (partial.type === 'FULL') {
+        transactions.push(partial.value)
+      } else if (currResponseIndex >= responseTransactions.length) {
+        // did not respond with enough transactions
+        return { ok: false }
+      } else {
+        const next = new Transaction(responseTransactions[currResponseIndex])
+        if (!next.hash().equals(partial.value)) {
+          // hashes are mismatched
+          return { ok: false }
+        }
+        transactions.push(next)
+        currResponseIndex++
+      }
+    }
+
+    return { ok: true, transactions }
   }
 
-  private handleNewBlockV2Message(peer: Peer, message: NewBlockV2Message) {
-    this.logger.debug(`Received unimplemented message ${message.type}`)
+  private async onNewBlockTransactions(peer: Peer, message: GetBlockTransactionsResponse) {
+    const block = this.blockFetcher.receivedBlockTransactions(message)
```

Comment:
I think maybe we should check `alreadyHaveBlock` here too

Reply:
I don't think it's necessary, unless there's an edge case I'm not thinking of. `receivedBlockTransactions` will return null for the block and drop the message if the block state isn't `TRANSACTION_REQUEST_IN_FLIGHT`, and we should be changing the state to `PROCESSING_FULL_BLOCK` in `onNewFullBlock`, or removing the block if it enters the chain, is marked as invalid, or is marked as an orphan.

Actually, this made me realize we don't remove blocks from the fetcher when marking them as an orphan, so I'll do that